File: apps/backend/app/schemas/chapter.py
from pydantic import BaseModel
from typing import Optional, List, Dict, Any
from datetime import datetime
import uuid


class ChapterCreate(BaseModel):
    title: str
    outline_node_id: Optional[uuid.UUID] = None
    content: str = ""
    sort_order: int = 0


class ChapterUpdate(BaseModel):
    title: Optional[str] = None
    content: Optional[str] = None
    manuscript_raw_snapshot: Optional[str] = None
    sort_order: Optional[int] = None
    status: Optional[str] = None


class ChapterOut(BaseModel):
    id: uuid.UUID
    project_id: uuid.UUID
    outline_node_id: Optional[uuid.UUID]
    title: str
    content: str
    manuscript_raw_snapshot: Optional[str] = None
    word_count: int
    sort_order: int
    status: str
    last_quality_score: Optional[float]
    last_quality_report: Optional[Dict[str, Any]] = None
    quality_checked_at: Optional[datetime] = None
    created_at: datetime
    updated_at: Optional[datetime]
    version: int = 1
    extra: Optional[Dict[str, Any]] = None

    class Config:
        from_attributes = True


class ChapterVersionOut(BaseModel):
    id: uuid.UUID
    chapter_id: uuid.UUID
    word_count: Optional[int]
    note: Optional[str]
    is_auto: bool
    created_at: datetime

    class Config:
        from_attributes = True


class ChapterVersionDetailOut(ChapterVersionOut):
    """单条版本详情（含正文 HTML），用于历史预览与恢复"""
    content: str

    class Config:
        from_attributes = True


class ChapterVersionTimelineItemOut(BaseModel):
    """项目维度：章节版本快照列表项（不含正文，用于时间线）"""

    id: uuid.UUID
    chapter_id: uuid.UUID
    chapter_title: str
    chapter_sort_order: int
    word_count: Optional[int]
    note: Optional[str]
    is_auto: bool
    created_at: datetime

    class Config:
        from_attributes = True

File: apps/client/src/types/chapter.ts
// ── Outline ───────────────────────────────────────────
export interface OutlineNode {
  id: string
  project_id: string
  parent_id?: string
  node_type: 'volume' | 'arc' | 'chapter_plan'
  title: string
  summary?: string
  hook?: string
  highlight?: string
  conflict?: string
  sort_order: number
  expected_words?: number
  reader_hook_score?: number
  extra: Record<string, any>
  children: OutlineNode[]
  created_at: string
  updated_at?: string
  // v2 新增字段
  storyline_ids?: string[]
  involved_character_ids?: string[]
  key_item_ids?: string[]
  key_skill_ids?: string[]
  emotional_tone?: string
  pacing?: string
  power_milestone?: string
  foreshadows_laid?: Array<{ id?: string; description: string }>
  foreshadows_resolved?: Array<{ id?: string; description: string }>
  phase?: string
  // P2 三层调度新增
  character_screen_time?: Record<string, number>
  pov_character_id?: string
}

// ── Quality ───────────────────────────────────────────
export interface QualityReport {
  overall_score: number
  dimensions: Record<string, {
    score: number
    status: 'excellent' | 'pass' | 'warning' | 'fail'
    comment: string
  }>
  issues: Array<{ type: string; description: string }>
  suggestions: string[]
  summary: string
}

// ── Chapter ───────────────────────────────────────────
export interface Chapter {
  id: string
  project_id: string
  outline_node_id?: string
  title: string
  content: string
  /** 最近一次 AI 返回的完整纯文本（含稿末索引），与 content 分离，用于「原文」对照 */
  manuscript_raw_snapshot?: string | null
  word_count: number
  sort_order: number
  status: 'draft' | 'writing' | 'done' | 'reviewed'
  last_quality_score?: number
  last_quality_report?: QualityReport
  quality_checked_at?: string
  created_at: string
  updated_at?: string
}

export interface ChapterVersion {
  id: string
  chapter_id: string
  word_count?: number
  note?: string
  is_auto: boolean
  created_at: string
}

export interface ChapterVersionDetail extends ChapterVersion {
  content: string
}

// ── ChapterIndex ──────────────────────────────────────
export interface ChapterIndex {
  id: string
  project_id: string
  chapter_id: string
  chapter_number: number
  story_day?: string
  core_events: Array<Record<string, unknown> | string>
  first_appearances: Array<Record<string, unknown>>
  actual_foreshadows_laid: Array<Record<string, unknown>>
  actual_foreshadows_resolved: Array<Record<string, unknown>>
  ending_hook?: string
  hook_strength: number
  continuity_notes: Array<Record<string, unknown> | string>
  created_at: string
  updated_at?: string
}

// ── Quality Debt ──────────────────────────────────────
export interface QualityDebt {
  id: string
  project_id: string
  /** 删章或归档后可能为空，仍可用 source_chapter_number 定位 */
  chapter_id?: string | null
  source_chapter_number: number
  issue_type: string
  severity: 'critical' | 'high' | 'medium' | 'low' | string
  status: 'pending' | 'resolved' | 'dismissed'
  summary: string
  suggested_fix?: string
  /** 作者手动记录的修复说明，会一并注入定向 AI 修复提示词 */
  author_notes?: string
  created_at: string
  updated_at?: string
}

// ── CharacterChangeLog ────────────────────────────────
export interface ChangeItem {
  field:  string
  label:  string
  before: string | null
  after:  string | null
}

export interface CharacterChangeLog {
  id:             string
  project_id:     string
  character_id:   string
  character_name: string
  chapter_id:     string | null
  chapter_number: string | null
  chapter_title:  string | null
  source:         'debrief' | 'manual' | 'bootstrap'
  summary:        string | null
  changes:        ChangeItem[]
  created_at:     string
}
